src/GraphQL/Validation/DocumentValidator.cs

```diff
@@ -13,10 +13,11 @@ namespace GraphQL.Validation
     public interface IDocumentValidator
     {
         /// <inheritdoc cref="IDocumentValidator"/>
-        Task<IValidationResult> ValidateAsync(
+        Task<(IValidationResult validationResult, Variables variables)> ValidateAsync(
```

Comment:
@Shane32 I still doubt this way of returning values. `out` does not work for async methods.

Reply:
To be honest, I generally doubt that an async/Task is needed here. I would generally make the method synchronous.

Reply:
Hmm.  I'd tend to agree, given the rest of the design of the validation code.  However, could this be used to asynchronously retrieve some information from a database prior to the validation?  Let's say someone wanted the validation code to be based on the permission set of the authenticated user (JWT/cookie/querystring/etc).  And assume that an async call is necessary for some reason to retrieve the permissions, whether it's from a redis cache or a database.  Would this be how it's done?  Is there another way to do it?  I think this is a single `Task` - a single allocation - per request, and as such, I'd probably leave it as-is.

Reply:
I guess maybe a few allocations - one for each validation rule.

Reply:
I understand, I just grumbled a little about the asynchronous code, which is 99% synchronous.